### ISSUE TITLE
Allow snake_case in the parser, make formatter prefer it over camelCase

### ISSUE THREAD

**Issue by HajagosNorbert** [roc-lang/roc#7214]
- Make the parser accept underscores for identifiers, even in words that begin with an uppercase (Tag names, Type names for examle)
- Change the formatter to enforce snake_case where we are using camelCase now (variable names for example). That is to automatically convert camelCase lowercase names into snake_case by replacing uppercase letters with underscore followed by lowercase letter, and then automatically combine consecutive underscores into 1 underscore so names never have 2 underscores in a row)
- Compiler warnings aren't needed for using camelCase just yet, both styles are accepted.

**Comment by gamebox** [comment]
If I were interested in contributing here, should I first join the Zulip (which I haven't used in maybe 5 years)?

**Comment by gamebox** [comment]
The first question I have as a beginner here is:  Would we like me to start by adding some unit tests to `crates/compiler/parse/src/ident.rust`?  I see no tests for this module and without having all the invariant testing in place, I would feel less confident moving ahead.

**Comment by gamebox** [comment]
I could start by just adding a range of tests for all invariants for `parseIdent`.

**Comment by lukewilliamboswell** [comment]
@joshuawarner32 is our resident parsing expert 😃 he may have ideas here. 

Tests sound good to me, but not sure if they're already covered in `cargo test -p test_syntax` over at `crates/compiler/test_syntax`

**Comment by joshuawarner32** [comment]
Yeah, unit tests in somewhere in the parse crate makes sense in this case, given we're talking about a relatively isolated bit of logic.

For any non-trivial tests (e.g. tests involving parsing an expression), I've been trying to standardize all of those to be part of the `test_snapshots` target in `test_syntax`, as @lukewilliamboswell pointed out. I'd recommend adding a few tests there and selecting a subset of the existing tests to pre-emptively convert to use either underscores or camelCase, just so we have a variety of coverage there.

**Comment by gamebox** [comment]
Cool, that follows my intuition as well.

Quick follow up, I noticed that when we parse `Ident::OpaqueRef` we keep the `@` sigil in the result string.  And given the assert I see in `chomp_opaque_ref` this seems important.  What's the motivation there?
